File: docs/experiments/03-compile-optimising-codegen/compile_optimising_codegen.py
#!/usr/bin/env python3
"""Experiment 3 -- Optimising codegen: CSE, folding, DCE, LICM.

Runs the same bodies as ``compile_optimising_codegen.m``.  See ``README.md``.

    python3 compile_optimising_codegen.py

WHAT THIS COLUMN IS FOR.  The experiment is about a bytecode compiler's
optimiser, which Python does not have an equivalent of -- CPython's peephole
optimiser does not do CSE or LICM across a loop.  These rows are therefore the
scalar-loop baseline the CAS optimisers are trying to beat, not a competing
optimiser.  **numba is not installed on this host**; with it, the first four
rows would be a genuine optimiser comparison.

The last row IS vectorisable and is a fair NumPy result.
"""

import time

import numpy as np


def bench(label, fn, reps=3):
    """One untimed warm-up, then the MINIMUM of `reps` timed runs."""
    fn()
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        fn()
        ts.append(time.perf_counter() - t0)
    print("%-52s%s ms" % (label, round(1000.0 * min(ts), 3)))


def check(label, value):
    print("%-52scheck = %s" % (label, value))


def horner(x):
    """A degree-40 Horner chain with constant coefficients."""
    s = 1.0
    for k in range(1, 41):
        s = s * x + (1.0 + 1.0 / (k + 1.0))
    return s


def cse(x, y):
    """A repeated subexpression: x*y appears twice."""
    import math
    return x * y + math.sin(x * y)


def licm(x, n):
    """A loop-invariant subexpression inside a while loop."""
    import math
    s, k = 0.0, 0
    while k < n:
        s = s + x * math.sqrt(2.0) + math.sin(1.0)
        k += 1
    return s


def newt(a):
    """Newton's method -- the micro-benchmark the codegen work was tuned on."""
    x, k = 1.0, 0
    while k < 20:
        x = x - (x * x - a) / (2.0 * x)
        k += 1
    return x


def main():
    print("Experiment 3 -- optimising codegen")
    print("")
    bench("Horner, degree 40, 10^5 calls (CPython)",
          lambda: [horner(0.5) for _ in range(10 ** 5)], reps=1)
    check("Horner, degree 40", horner(0.5))

    bench("x*y + sin(x*y), 10^5 calls (CPython)",
          lambda: [cse(0.5, 0.25) for _ in range(10 ** 5)], reps=1)
    check("x*y + sin(x*y)", cse(0.5, 0.25))

    bench("loop-invariant body, 10^6 iterations (CPython)",
          lambda: licm(0.5, 10 ** 6), reps=1)
    check("loop-invariant body", licm(0.5, 10))

    bench("Newton, 20 iterations, 10^5 calls (CPython)",
          lambda: [newt(2.0) for _ in range(10 ** 5)], reps=1)
    check("Newton (sqrt 2)", newt(2.0))

    x = np.linspace(0.0, 1.0, 10 ** 6)
    bench("degree-5 polynomial over 10^6 points (NumPy)",
          lambda: 1.0 + 2.0 * x + 3.0 * x ** 2 + 4.0 * x ** 3
                  + 5.0 * x ** 4 + 6.0 * x ** 5)


if __name__ == "__main__":
    main()

File: docs/experiments/03-compile-optimising-codegen/compile_optimising_codegen.m
(* ==========================================================================
   Experiment 3 -- Optimising codegen: CSE, folding, DCE, LICM
   ==========================================================================

   Runs unmodified in Mathilda and in Mathematica 14. See README.md.

       Mathilda      ./Mathilda -file compile_optimising_codegen.m
       Mathematica   wolframscript -file compile_optimising_codegen.m
       Python        python3 compile_optimising_codegen.py

   WHAT IT MEASURES.  Bodies chosen so that a specific optimisation either
   fires or does not: a long Horner chain (constant folding and instruction
   count), a repeated subexpression (CSE), a loop with an invariant (LICM),
   and a dead branch (DCE).

   WHY THE CEILING IS LOW, stated up front because the numbers look
   unimpressive next to experiment 1's 234x: a bytecode VM's dispatch is
   already the dominant cost, so removing instructions helps only in
   proportion to how many are removed.  1.05-1.48x is what that is worth, and
   the experiment exists to establish the number rather than to guess it.
   ========================================================================== *)

SetAttributes[bench, HoldRest];

(* bench[label, expr] -- one untimed warm-up, then the MINIMUM of three timed
   runs.  The minimum, not the mean: we are measuring the cost of the work, and
   every source of noise on a loaded machine can only add. *)
bench[label_String, expr_] := Module[{ts},
  expr;
  ts = Table[First[AbsoluteTiming[expr]], {3}];
  Print[StringPadRight[label, 52], ToString[Round[1000. Min[ts], 0.001]], " ms"]
];

check[label_String, value_] :=
  Print[StringPadRight[label, 52], "check = ", value];


(* A degree-40 Horner chain: 40 multiply-adds, all of whose coefficients are
   constant expressions the folder can evaluate once. *)
horner = Compile[{{x, _Real}},
  Module[{s = 0., k = 0},
    s = 1.;
    Do[s = s x + (1. + 1./(k + 1.)), {k, 1, 40}];
    s]];

(* A repeated subexpression: x y appears twice, so CSE should compute it once. *)
cse = Compile[{{x, _Real}, {y, _Real}}, x y + Sin[x y]];

(* A loop-invariant subexpression inside a While. *)
licm = Compile[{{x, _Real}, {n, _Integer}},
  Module[{s = 0., k = 0},
    While[k < n, s = s + x Sqrt[2.] + Sin[1.]; k = k + 1]; s]];

(* Newton's method: the micro-benchmark the codegen work was tuned on. *)
newt = Compile[{{a, _Real}},
  Module[{x = 1., k = 0},
    While[k < 20, x = x - (x x - a)/(2. x); k = k + 1]; x]];

Print["Experiment 3 -- optimising codegen"];
Print[""];

bench["Horner, degree 40, 10^6 calls",  Do[horner[0.5], {10^6}]];
check["Horner, degree 40",              horner[0.5]];

bench["x y + Sin[x y], 10^6 calls (CSE)", Do[cse[0.5, 0.25], {10^6}]];
check["x y + Sin[x y]",                   cse[0.5, 0.25]];

bench["loop-invariant body, 10^6 iterations (LICM)", licm[0.5, 10^6]];
check["loop-invariant body",                          licm[0.5, 10]];

bench["Newton, 20 iterations, 10^5 calls", Do[newt[2.], {10^5}]];
check["Newton (sqrt 2)",                   newt[2.]];

bench["degree-5 polynomial over 10^6 points",
      Table[1. + 2. x + 3. x^2 + 4. x^3 + 5. x^4 + 6. x^5, {x, 0., 1., 1./(10^6 - 1)}]];
